lib/pure/bitops.nim

```diff
@@ -163,174 +164,208 @@ elif useICC_builtins:
-                                    builtin_popcnt32((x.uint64 shr 32'u64).cint ).int
+      when sizeof(x) <= 4: builtin_popcnt32(x.cint).int
+      elif arch64:         builtin_popcnt64(x.uint64).int
+      else:                builtin_popcnt32((x.uint64 and 0xFFFFFFFF'u64).cint ).int +
+                          builtin_popcnt32((x.uint64 shr 32'u64).cint ).int
     else:
-      when sizeof(x) <= 4: result = countSetBits_nim(x.uint32)
-      else:                result = countSetBits_nim(x.uint64)
+      when sizeof(x) <= 4: oneBitsNim(x.uint32)
+      else:                oneBitsNim(x.uint64)
+
+func countSetBits*(x: SomeInteger): int {.deprecated: "oneBits".} =
+  oneBits(x.uint64)
 
-proc popcount*(x: SomeInteger): int {.inline, nosideeffect.} =
+func popcount*(x: SomeInteger): int {.deprecated: "oneBits".}=
   ## Alias for for countSetBits (Hamming weight.)
-  result = countSetBits(x)
+  oneBits(x.uint64)
 
-proc parityBits*(x: SomeInteger): int {.inline, nosideeffect.} =
+func parityBits*(x: SomeUnsignedInt): int =
   ## Calculate the bit parity in integer. If number of 1-bit
   ## is odd parity is 1, otherwise 0.
+  ##
+  ## Example:
+  ## doAssert parityBits(0b00000001'u8) == 1
   # Can be used a base if creating ASM version.
   # https://stackoverflow.com/questions/21617970/how-to-check-if-value-has-even-parity-of-bits-or-odd
   when nimvm:
-    when sizeof(x) <= 4: result = parity_impl(x.uint32)
-    else:                result = parity_impl(x.uint64)
+    when sizeof(x) <= 4: parityImpl(x.uint32)
+    else:                parityImpl(x.uint64)
   else:
     when useGCC_builtins:
-      when sizeof(x) <= 4: result = builtin_parity(x.uint32).int
-      else:                result = builtin_parityll(x.uint64).int
+      when sizeof(x) <= 4: builtin_parity(x.uint32).int
+      else:                builtin_parityll(x.uint64).int
     else:
-      when sizeof(x) <= 4: result = parity_impl(x.uint32)
-      else:                result = parity_impl(x.uint64)
+      when sizeof(x) <= 4: parityImpl(x.uint32)
+      else:                parityImpl(x.uint64)
+
+func parityBits*(x: SomeSignedInt): int {.deprecated.} =
+  parityBits(x.uint64)
 
-proc firstSetBit*(x: SomeInteger): int {.inline, nosideeffect.} =
+func firstOneBit*(x: SomeUnsignedInt, maybeZero = true): int =
   ## Returns the 1-based index of the least significant set bit of x.
-  ## If `x` is zero, when ``noUndefinedBitOpts`` is set, result is 0,
-  ## otherwise result is undefined.
-  # GCC builtin 'builtin_ffs' already handle zero input.
+  ## If `x` is zero and `maybeZero` is true, result is 0
+  ## If `x` is zero and `maybeZero` is false, result is undefined
+  ##
+  ## Example:
+  ## doAssert firstOneBit(0b00000010'u8) == 2
+  ##
   when nimvm:
-    when noUndefined:
-      if x == 0:
-        return 0
-    when sizeof(x) <= 4: result = firstSetBit_nim(x.uint32)
-    else:                result = firstSetBit_nim(x.uint64)
+    if maybeZero and x == 0: 0
```

Comment:
With a static bool, GCC/LLVM can optimize this + the following logic into LZCNT

Reply:
it can also optimize it without a static bool